clusterloader2/api/validation/validation.go

```diff
@@ -0,0 +1,148 @@
+limitations under the License.
+*/
+
+package validation
+
+import (
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/perf-tests/clusterloader2/api"
+)
+
+// VerifyConfig does simple validation for config and all included structures.
+// Returns error is any condition is not satisfied.
+func VerifyConfig(c *api.Config) field.ErrorList {
+	allErrs := field.ErrorList{}
+	fldPath := field.NewPath(".")
+	if c.AutomanagedNamespaces < 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath, "automanagedNamespaces", "must be non-negative"))
+	}
+	for i, tuningSet := range c.TuningSets {
+		allErrs = append(allErrs, verifyTuningSet(&tuningSet, fldPath.Child("tuningSets").Index(i))...)
+	}
+	for i, step := range c.Steps {
+		allErrs = append(allErrs, verifyStep(&step, fldPath.Child("steps").Index(i))...)
+	}
+	return allErrs
+}
+
+// verifyStep requires to either Phases or Measurements be non-empty,
+// but never both at the same time.
+func verifyStep(s *api.Step, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	for i, phase := range s.Phases {
+		allErrs = append(allErrs, verifyPhase(&phase, fldPath.Child("phases").Index(i))...)
+	}
+	for i, measurement := range s.Measurements {
+		allErrs = append(allErrs, verifyMeasurement(&measurement, fldPath.Child("measurements").Index(i))...)
+	}
+	if (len(s.Phases) == 0) == (len(s.Measurements) == 0) {
+		allErrs = append(allErrs, field.Invalid(fldPath, "phases, measurements", "exactly one should non-empty"))
+	}
+	return allErrs
+}
+
+// verifyPhase requires ReplicasPerNamespace to have positive value.
+func verifyPhase(p *api.Phase, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if p.NamespaceRange != nil {
+		allErrs = append(allErrs, verifyNamespaceRange(p.NamespaceRange, fldPath.Child("namespaceRange"))...)
+	}
+	if p.ReplicasPerNamespace < 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath, "replicasPerNamespace", "must be non-negative"))
+	}
+	for i, object := range p.ObjectBundle {
+		allErrs = append(allErrs, verifyObject(&object, fldPath.Child("objectBundle").Index(i))...)
+	}
+	return allErrs
+}
+
+func verifyObject(o *api.Object, fldPath *field.Path) field.ErrorList {
+	return verifyObjectType(&o.ObjectType, fldPath.Child("objectType"))
+}
+
+// verifyNamespaceRange requires Min and Max to have positive values
+// and Max to be greater than Min.
+func verifyNamespaceRange(nr *api.NamespaceRange, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if nr.Min < 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath, "min", "must be non-negative"))
+	}
+	if nr.Max < 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath, "max", "must be non-negative"))
+	}
+	if nr.Max < nr.Min {
+		allErrs = append(allErrs, field.Invalid(fldPath, "min, max", "Max must be greater than Min"))
+	}
+	return allErrs
+}
+
+func verifyObjectType(ot *api.ObjectType, fldPath *field.Path) field.ErrorList {
+	return field.ErrorList{}
```

Comment:
That does unnecessary memory allocation.

return nil

Reply:
done